File: android/sdk/src/main/java/com/viant/forgeandroid/ui/AssignmentPickerRenderer.kt
package com.viant.forgeandroid.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.Checkbox
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.viant.forgeandroid.runtime.AssignmentPickerSpec
import com.viant.forgeandroid.runtime.ContainerDef
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.SelectorUtil
import com.viant.forgeandroid.runtime.WindowContext

@Composable
internal fun AssignmentPickerRenderer(runtime: ForgeRuntime, window: WindowContext, container: ContainerDef) {
    val spec = container.assignmentPicker ?: return
    val availableContext = window.contextOrNull(spec.availableDataSourceRef) ?: return
    val assignedContext = window.contextOrNull(spec.assignedDataSourceRef) ?: return
    val available by availableContext.collection.flow.collectAsState(initial = availableContext.collection.peek())
    val assigned by assignedContext.collection.flow.collectAsState(initial = assignedContext.collection.peek())
    var availableSelection by remember { mutableStateOf(emptySet<String>()) }
    var assignedSelection by remember { mutableStateOf(emptySet<String>()) }
    val assignedIds = assigned.mapNotNull { assignmentIdentity(it, spec) }.toSet()
    val assignable = available.filter { assignmentIdentity(it, spec)?.let { id -> id !in assignedIds } == true }

    LaunchedEffect(spec.availableDataSourceRef, spec.assignedDataSourceRef) {
        if (availableContext.dataSource.autoFetch != false) availableContext.fetchCollection()
        if (assignedContext.dataSource.autoFetch != false) assignedContext.fetchCollection()
    }
    LaunchedEffect(assignable) { availableSelection = availableSelection.intersect(assignable.mapNotNull { assignmentIdentity(it, spec) }.toSet()) }
    LaunchedEffect(assigned) { assignedSelection = assignedSelection.intersect(assigned.mapNotNull { assignmentIdentity(it, spec) }.toSet()) }

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(12.dp), modifier = Modifier.fillMaxWidth()) {
            AssignmentColumn("Available", assignable, spec, availableSelection, Modifier.weight(1f)) { id ->
                availableSelection = toggleAssignment(availableSelection, id, spec.allowMultiple != false)
            }
            AssignmentColumn("Assigned", assigned, spec, assignedSelection, Modifier.weight(1f)) { id ->
                assignedSelection = toggleAssignment(assignedSelection, id, spec.allowMultiple != false)
            }
        }
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            spec.assign?.let { command ->
                MutationCommandButton(
                    runtime, window, availableContext, command,
                    labelOverride = command.label ?: "Assign",
                    extras = mapOf("selectedRows" to assignable.filter { assignmentIdentity(it, spec)?.let(availableSelection::contains) == true }),
                    externallyDisabled = availableSelection.isEmpty()
                )
            }
            spec.unassign?.let { command ->
                MutationCommandButton(
                    runtime, window, assignedContext, command,
                    labelOverride = command.label ?: "Unassign",
                    extras = mapOf("selectedRows" to assigned.filter { assignmentIdentity(it, spec)?.let(assignedSelection::contains) == true }),
                    externallyDisabled = assignedSelection.isEmpty()
                )
            }
        }
    }
}

@Composable
private fun AssignmentColumn(
    title: String,
    rows: List<Map<String, Any?>>,
    spec: AssignmentPickerSpec,
    selection: Set<String>,
    modifier: Modifier,
    onToggle: (String) -> Unit
) {
    Column(
        verticalArrangement = Arrangement.spacedBy(4.dp),
        modifier = modifier.background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f), RoundedCornerShape(10.dp)).padding(10.dp)
    ) {
        Text(title, style = MaterialTheme.typography.titleMedium)
        if (rows.isEmpty()) Text("None", style = MaterialTheme.typography.bodySmall)
        rows.forEach { row ->
            val identity = assignmentIdentity(row, spec)
            Row(modifier = Modifier.fillMaxWidth()) {
                Checkbox(checked = identity != null && identity in selection, enabled = identity != null, onCheckedChange = { if (identity != null) onToggle(identity) })
                Text(assignmentLabel(row, spec), modifier = Modifier.padding(top = 12.dp))
            }
        }
    }
}

private fun assignmentIdentity(row: Map<String, Any?>, spec: AssignmentPickerSpec): String? {
    val fields = spec.identityFields.ifEmpty { listOf("id") }
    val values = fields.map { row[it]?.toString()?.takeIf(String::isNotBlank) ?: return null }
    return values.joinToString("\u001f")
}

private fun assignmentLabel(row: Map<String, Any?>, spec: AssignmentPickerSpec): String =
    SelectorUtil.resolve(row, spec.labelField ?: "label")?.toString()?.takeIf(String::isNotBlank)
        ?: assignmentIdentity(row, spec)
        ?: "Item"

private fun toggleAssignment(selection: Set<String>, id: String, multiple: Boolean): Set<String> = when {
    id in selection -> selection - id
    !multiple -> setOf(id)
    else -> selection + id
}

File: ios/Sources/ForgeIOSUI/AssignmentPickerRenderer.swift
import SwiftUI
import ForgeIOSRuntime

struct AssignmentPickerRenderer: View {
    let runtime: ForgeRuntime
    let window: WindowContext
    let container: ContainerDef

    @State private var available: [[String: JSONValue]] = []
    @State private var assigned: [[String: JSONValue]] = []
    @State private var availableSelection: Set<String> = []
    @State private var assignedSelection: Set<String> = []

    private var spec: AssignmentPickerSpec { container.assignmentPicker! }
    private var identityFields: [String] { spec.identityFields?.isEmpty == false ? spec.identityFields! : ["id"] }
    private var assignedIDs: Set<String> { Set(assigned.compactMap(identity)) }
    private var assignable: [[String: JSONValue]] { available.filter { row in identity(row).map { !assignedIDs.contains($0) } ?? false } }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(alignment: .top, spacing: 12) {
                pickerColumn(title: "Available", rows: assignable, selection: $availableSelection)
                pickerColumn(title: "Assigned", rows: assigned, selection: $assignedSelection)
            }
            HStack(spacing: 8) {
                if let command = spec.assign {
                    MutationCommandButton(
                        runtime: runtime,
                        window: window,
                        sourceDataSourceRef: spec.availableDataSourceRef,
                        command: command,
                        labelOverride: command.label ?? "Assign",
                        extras: ["selectedRows": .array(selectedRows(assignable, ids: availableSelection).map(JSONValue.object))],
                        externallyDisabled: availableSelection.isEmpty
                    )
                }
                if let command = spec.unassign {
                    MutationCommandButton(
                        runtime: runtime,
                        window: window,
                        sourceDataSourceRef: spec.assignedDataSourceRef,
                        command: command,
                        labelOverride: command.label ?? "Unassign",
                        extras: ["selectedRows": .array(selectedRows(assigned, ids: assignedSelection).map(JSONValue.object))],
                        externallyDisabled: assignedSelection.isEmpty
                    )
                }
            }
        }
        .task(id: "\(window.windowID)#\(spec.availableDataSourceRef)#\(spec.assignedDataSourceRef)") {
            await observeCollections()
        }
        .onChange(of: assignable.compactMap(identity)) { _, ids in availableSelection.formIntersection(Set(ids)) }
        .onChange(of: assigned.compactMap(identity)) { _, ids in assignedSelection.formIntersection(Set(ids)) }
        .accessibilityIdentifier("forge-assignment-picker")
    }

    private func pickerColumn(title: String, rows: [[String: JSONValue]], selection: Binding<Set<String>>) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(title).font(.headline)
            if rows.isEmpty {
                Text("None").font(.caption).foregroundStyle(.secondary)
            } else {
                ForEach(Array(rows.enumerated()), id: \.offset) { index, row in
                    let rowID = identity(row) ?? "invalid-\(index)"
                    Button {
                        if spec.allowMultiple == false {
                            selection.wrappedValue = selection.wrappedValue.contains(rowID) ? [] : [rowID]
                        } else if selection.wrappedValue.contains(rowID) {
                            selection.wrappedValue.remove(rowID)
                        } else {
                            selection.wrappedValue.insert(rowID)
                        }
                    } label: {
                        HStack {
                            Image(systemName: selection.wrappedValue.contains(rowID) ? "checkmark.square.fill" : "square")
                            Text(label(row))
                            Spacer(minLength: 0)
                        }
                    }
                    .buttonStyle(.plain)
                    .disabled(identity(row) == nil)
                    .accessibilityValue(selection.wrappedValue.contains(rowID) ? "Selected" : "Not selected")
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .topLeading)
        .padding(10)
        .background(.secondary.opacity(0.06), in: RoundedRectangle(cornerRadius: 10))
    }

    @MainActor
    private func observeCollections() async {
        available = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: spec.availableDataSourceRef)
        assigned = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: spec.assignedDataSourceRef)
        if available.isEmpty { await runtime.refreshDataSourceCollection(windowID: window.windowID, dataSourceRef: spec.availableDataSourceRef) }
        if assigned.isEmpty { await runtime.refreshDataSourceCollection(windowID: window.windowID, dataSourceRef: spec.assignedDataSourceRef) }
        available = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: spec.availableDataSourceRef)
        assigned = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: spec.assignedDataSourceRef)
        await withTaskGroup(of: Void.self) { group in
            group.addTask {
                let stream = await runtime.dataSourceCollectionUpdates(windowID: window.windowID, dataSourceRef: spec.availableDataSourceRef)
                for await rows in stream { await MainActor.run { available = rows } }
            }
            group.addTask {
                let stream = await runtime.dataSourceCollectionUpdates(windowID: window.windowID, dataSourceRef: spec.assignedDataSourceRef)
                for await rows in stream { await MainActor.run { assigned = rows } }
            }
        }
    }

    private func identity(_ row: [String: JSONValue]) -> String? {
        let values = identityFields.compactMap { field -> String? in
            guard let value = row[field] else { return nil }
            switch value {
            case .string(let value): return value.isEmpty ? nil : value
            case .number(let value): return value.rounded() == value ? String(Int(value)) : String(value)
            default: return nil
            }
        }
        return values.count == identityFields.count ? values.joined(separator: "\u{001f}") : nil
    }

    private func label(_ row: [String: JSONValue]) -> String {
        let field = spec.labelField ?? "label"
        return assignmentJSONValue(SelectorUtil.resolve(row.mapValues(assignmentAnyValue), selector: field))?.stringValue
            ?? identity(row)
            ?? "Item"
    }

    private func selectedRows(_ rows: [[String: JSONValue]], ids: Set<String>) -> [[String: JSONValue]] {
        rows.filter { identity($0).map(ids.contains) == true }
    }
}

private func assignmentAnyValue(_ value: JSONValue) -> Any {
    switch value {
    case .string(let value): return value
    case .number(let value): return value
    case .bool(let value): return value
    case .array(let values): return values.map(assignmentAnyValue)
    case .object(let values): return values.mapValues(assignmentAnyValue)
    case .null: return NSNull()
    }
}

private func assignmentJSONValue(_ value: Any?) -> JSONValue? {
    switch value {
    case let value as String: return .string(value)
    case let value as Double: return .number(value)
    case let value as Int: return .number(Double(value))
    case let value as Bool: return .bool(value)
    default: return nil
    }
}
